### PR TITLE
Standarize conditions across component operators

### PR DESCRIPTION
**Pull request by rthallisey**
Conditions are how component operators will report status to the HCO.  We want to pass that status on to the user and use a readiness probe to inform OLM of sensitive lifecycle operations.

Conditions will be reported on the component operator CRs.

### CODE REVIEW

**Comment by rthallisey** [comment]
cc @mhrivnak @jarrpa

File: docs/conditions.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Conditions are..
+	   _the latest available observations of an object's state. They are
+	   an extension mechanism intended to be used when the details of an
+	   observation are not a priori known or would not apply to all
+	   instances of a given Kind._
+
+The HCO’s CR is a representation of the underlying component operators.  If the
+object exists, then all components exist.  If the object doesn’t exist, all
+components don’t exist.  However, the CR existence doesn’t help us with the
+state where the operators exist, but are they healthy?  This is where conditions
```
Comment (djzager) [inline docs/conditions.md]: I believe it's worth pointing out the `Reason` and `Message` (and maybe the whole Condition struct) to make it clear that we should be giving a reason and message for transitions into (and out of) the different ConditionTypes.

Copying the [CVO types](https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go):

```
type ClusterOperatorStatusCondition struct {
	// type specifies the state of the operator's reconciliation functionality.
	Type ClusterStatusConditionType `json:"type"`

	// status of the condition, one of True, False, Unknown.
	Status ConditionStatus `json:"status"`

	// lastTransitionTime is the time of the last update to the current status object.
	LastTransitionTime metav1.Time `json:"lastTransitionTime"`

	// reason is the reason for the condition's last transition.  Reasons are CamelCase
	Reason string `json:"reason,omitempty"`

	// message provides additional information about the current condition.
	// This is only to be consumed by humans.
	Message string `json:"message,omitempty"`
}
```

File: docs/conditions.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# CR Conditions and Readiness Probe
+Conditions are..
```
Comment (rmohr) [inline docs/conditions.md]: They are also backward compatible. State machine changes (e.g. via status) are not.

File: docs/conditions.md (hunk 1 above)
Comment (rmohr) [reply docs/conditions.md]: :+1: Let's explicitly point out that the conditions are the same condition type like in k8s (and CVO).

Comment (rmohr) [reply docs/conditions.md]: Oh, and let's maybe also standardize some of the `Reson` strings: e.g. `Healing` or `Updating`. I guess we can find some in the okd repo.

Comment (djzager) [reply docs/conditions.md]: Just keep in mind that, assuming we follow the pattern set by CVO, there would be at least 3 ConditionTypes `Progressing|Available|Degraded` that can be `True|False|Unkown`. It may be difficult to standardize on the reasons for all of those but I would be interested in the proposal.

Comment (rthallisey) [reply docs/conditions.md]: I only talked about below `True and False` and left out `Unknown` because I wasn't sure how to write the `Meaning` section.

For instance, if you're `OperatorAvailable = true` `OperatorProgressing = unknown` and `OperatorDegraded = false`, what definitive meaning can we gather?  Does `Unknown` imply the operator doesn't know what's going on?  If that's the case then `Unknown` could be the "yellow warning light" to call for help?   Maybe that can be useful during failed upgrades...

Comment (djzager) [reply docs/conditions.md]: My comment about "Just keep in mind" was more directed at the standardizing on `Reason` string.

I don't have an immediate understanding of a scenario where the `Progressing` Condition would be `Unknown`...maybe when we first create the conditions slice? But I agree this proposal should focus on what `True|False` mean.

File: docs/conditions.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#### OperatorDegraded
+```
+	OperatorDegraded ClusterStatusConditionType = "Degraded"
+```
+Degraded indicates that the component operator is not functioning completely.
+An example of a degraded state would be if there should be 5 copies of the
+component running but only 4 are running. It may still be available, but it is
+degraded.
+
+#### Condition Matrix
```
Comment (phoracek) [inline docs/conditions.md]: Maybe mention what is the default value of (missing) condition.

File: docs/conditions.md (hunk 1 above)
Comment (rmohr) [reply docs/conditions.md]: > Just keep in mind that, assuming we follow the pattern set by CVO, there would be at least 3 ConditionTypes `Progressing|Available|Degraded` that can be `True|False|Unkown`. It may be difficult to standardize on the reasons for all of those but I would be interested in the proposal.

Yes you are right. Let me rephrase, maybe some suggestions would be good. For instance k8s uses `FailedCreate`, `FailedUpdate`, to indicate REST issues. But I guess it is more important to have the conditions standardized.

File: docs/conditions.md (hunk 3 above)
Comment (rthallisey) [reply docs/conditions.md]: What do you mean by 
> default value of (missing) condition
>
Are you asking what will happen if the HCO notices a CR hasn't implemented conditions or a particular condition?

Comment (phoracek) [reply docs/conditions.md]: Every CR is initially without any state - conditions. If HCO sees this initial state, how does it evaluates condition value? True, False or Unknown.

Comment (rthallisey) [reply docs/conditions.md]: If the HCO sees the CR with no value for the conditions, then the HCO will assume the Operator's condition is `OperatorAvailable = false`, `OperatorProgressing = unknown` and `OperatorDegraded = unknown`.  This is a good place for `unknown`

File: docs/conditions.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+See this [issue](https://github.com/operator-framework/operator-lifecycle-manager/issues/922) for why we only want to report a readiness probe on the HCO
+instead of on all component operators.
+
+## Reason
+`Reason` is _a one-word CamelCase reason for the condition's last transition_.
+
+We'll use a series of lifecycle inspired prefixes paired with postfixes to
+standardize values for `Reason`.
+
+|         | -Failed  | -Succeeded | -Invalid | -InProgress |
```
Comment (djzager) [inline docs/conditions.md]: Table isn't rendering, making this hard to review.

**Comment by slintes** [comment]
Just crossed my mind: how do we deal with operators running in HA mode, so multiple instances? How should HCO find the leader? Maybe with another condition?

File: docs/conditions.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+| :------------- |:-------------:|:-----:|:-----:|:-----:|
+| Install- | InstallFailed | InstallSucceeded | InstallInvalid | InstallInProgress |
```
Comment (djzager) [inline docs/conditions.md]: It is not clear to me which Reason you would use when transitioning to `Available: true`. I'm not really sure what standardizing on the `Reason` field really buys us at this point. IMO prescribing that operators populate the Reason is sufficient.

File: docs/conditions.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+`Message` is a _human-readable message indicating details about last transition_.
+
+Explain why your CR has `Reason`.
+
+
+## HCO conditions
+It's important to point out that the `ConditionTypes` on the HCO don't represent
+the condition the HCO is in, rather the condition of the component CRs.
+
+The HCO will use the same `ConditionType`s and `Reason`s, but it will be the
```
Comment (djzager) [inline docs/conditions.md]: Again, the scenarios in which an operator could be `Degraded` and `Unavailable` can't be pinned down to one Reason. I think we should skip Reason for now.

File: docs/conditions.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+The HCO's `ConditionType`s will always represent the _worst_ `Status` and the
+_most recently viewed_ `Reason`.
+
+The _worst_ `Status` for each `ConditionType`:
+
+| Condition   | Status  |
+| :------------- |:-------------:|
+| OperatorAvailable | False |
+| OperatorProgressing | False |
```
Comment (djzager) [inline docs/conditions.md]: I think that if any operator is Progressing...then the HCO is progressing.

File: docs/conditions.md (hunk 6 above)
Comment (rmohr) [reply docs/conditions.md]: Works for me. Let's drop reasons for now.

**Comment by rthallisey** [comment]
Going to change the names of the conditions to replace `Operator` with `Application` to make it more clear what the condition refers to.   Does anyone object?

File: docs/conditions.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+Conditions are..
+	   _the latest available observations of an object's state. They are
+	   an extension mechanism intended to be used when the details of an
+	   observation are not a priori known or would not apply to all
+	   instances of a given Kind._
+
+Kubernetes conditions [documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).
+
+The HCO’s CR is a representation of the all the underlying component operators'
+state.  If the the HCO's CR exists, then all component CRs exist, and all
```
[the remark this answers is not in the document]
Comment (rthallisey) [reply docs/conditions.md]: That's true, but I think this is a good example.  I'll change the language around to be more accurate.

File: docs/conditions.md (hunk 7 above)
Comment (rthallisey) [reply docs/conditions.md]: Do you think this should be renamed on the HCO?

**Comment by rthallisey** [comment]
@all I want to wrap up the design by tomorrow, Wednesday.  If you have any additional comments or if anything is unclear, please speak up and we can work to resolve it.

File: docs/conditions.md (hunk 7 above)
Comment (djzager) [reply docs/conditions.md]: This was more to point out that when you say "worst" for a given condition type...that doens't really work for the OperatorProgressing (or ApplicationProgressing). Since, what we want to communicate, is that if any of the components are progressing, then the HCO is also progressing.

This wasn't really about renaming anything. Just that the "worst" for progressing should be "True"

Comment (rthallisey) [reply docs/conditions.md]: I agree.  If Progressing is shown on the HCO as false and one operator has "True", it should be shown as "True" or that will be confusing.

"Worst" doesn't apply uniformly for all the conditions because the worst case is that we're failed and we're not doing anything about it.  OperatorProgressing is really an outlier here.

File: docs/conditions.md [hunk 9]
```diff
@@ -44,6 +44,13 @@ type ApplicationStatusCondition struct {
 }
 ```
 
+## Library
+We're going to use a shared library to provide the condition types to operators.
+This will ensure the code is not specific to any operator and it will allow
+products outside of CNV to also consume it.
+
+https://github.com/openshift-kni/operator-status
```
Comment (mhrivnak) [inline docs/conditions.md]: @markmc WDYT? We talked to operator-sdk about owning this, but they don't want to commit to it until they get past some other priorities. All operators in the CNV and OCS ecosystems will need to share these conditions, so openshift-kni seems like a reasonable org to be their home for now.

Comment (markmc) [reply docs/conditions.md]: I see openshift-kni as downstream of openshift, kubevirt, etc. - suggest doing this under github.com/openshift

File: docs/conditions.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+state.  In theory, if the HCO's CR exists, then all component CRs _should_
+exist, and all applications _should_ exist.  If the object doesn’t exist, then
+all component CRs _should not_ exist, and all applications _should not_ exist.
+However, the CR existence can only can tell us if the application should exist and
+doesn't help us observe the application's health. This is where the HCO and
+component operators will use conditions on their CRs to reflect the health of
+the underlying application.  Component operators store conditions that are
+watched by the HCO and the HCO will store conditions that reflect the
+[worst state](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/conditions.md#hco-conditions) of all component operator conditions.
+
```
Comment (markmc) [inline docs/conditions.md]: FWIW, in a drain controller design discussion this week, I saw Doug say:

> All the advice I've seen recently says that "conditions" are no longer the preferred way to express overall status, and that explicit fields should be used instead. The reason I've seen is that having a list of conditions makes it harder to make (or maybe keep) API promises than having regular fields on the CRD.

(I don't know have any further background on this line of thinking, and Doug is away this week. So again, just FWIW)

Comment (djzager) [reply docs/conditions.md]: Interesting that you mention that. So the [api-conventions doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) is slightly older than these issues talking about Status and Conditions:

- https://github.com/kubernetes/kubernetes/issues/50798
- https://github.com/kubernetes/kubernetes/issues/7856

There do seem to be some reasonable arguments on both sides of this discussion. At the same time, this is exactly what the Cluster Version Operator is doing. Without having a convincing argument to not use Conditions or clearly established alternative, I still think Conditions are the way to move forward.

Comment (rthallisey) [reply docs/conditions.md]: This might be the issue where it was discussed https://github.com/kubernetes-sigs/cluster-api/pull/1081 ?

Comment (rthallisey) [reply docs/conditions.md]: I think what's outlined in this doc follows the "outlook" approach written [here](https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-99667941).  Since this topic is long running discussion, I'm in favor of implementing this approach and following whatever the resolution of https://github.com/kubernetes/kubernetes/issues/7856 is at a later date.
